solr/core/src/java/org/apache/solr/core/NodeConfig.java

```diff
@@ -223,17 +221,23 @@ private NodeConfig(
 
     if (this.clusterSingletonPlugins != null
         && this.clusterSingletonPlugins.length > 0
-        && !ClusterPluginsSourceConfigurator.resolveClassName(this.clusterPluginsSource)
+        && !ClusterPluginsSourceConfigurator.resolveClassName()
             .equals(NodeConfigClusterPluginsSource.class.getName())) {
       throw new SolrException(
           ErrorCode.SERVER_ERROR,
-          "clusterSingleton section found in solr.xml but clusterPluginsSource is not NodeConfigClusterPluginsSource");
+          "clusterSingleton section found in solr.xml but the property "
+              + CONFIG_EDITING_DISABLED_ARG
+              + " is set to false. clusterSingleton plugins may only be declared in solr.xml with immutable configs.");
     }
 
     setupSharedLib();
     initModules();
   }
 
+  public static boolean isImmutableConfigSet() {
```

Comment:
My proposal is that this *not* be configSet specific.  So simply rename it to be more general.  Also FYI note that upstream there's a new sys-prop/env-var system and a dev list discussion.